File: java/lld/money-ledger/digital-wallet/TransferService.java
public class TransferService {

    public static class InsufficientFunds extends RuntimeException {}

    public static class TransferResult {
        public final String status = "OK";
        public final long amountCents;
        public TransferResult(long amountCents) { this.amountCents = amountCents; }
    }

    private final IdempotencyStore idempotency;

    public TransferService(IdempotencyStore idempotency) { this.idempotency = idempotency; }

    // Debits one ledger and credits another as a single atomic step. The two
    // ledgers are locked in a FIXED order (by wallet id) so a concurrent
    // transfer running the opposite direction can never deadlock against
    // this one.
    public TransferResult transfer(String fromId, Ledger fromLedger, String toId, Ledger toLedger, long amountCents, String idempotencyKey) {
        TransferResult cached = idempotency.get(idempotencyKey);
        if (cached != null) return cached;                   // retry — do nothing new

        boolean fromFirst = fromId.compareTo(toId) < 0;
        Ledger first = fromFirst ? fromLedger : toLedger;
        Ledger second = fromFirst ? toLedger : fromLedger;

        synchronized (first) {
            synchronized (second) {
                if (fromLedger.getBalance() < amountCents) throw new InsufficientFunds();
                fromLedger.append(LedgerEntry.Type.DEBIT, amountCents);
                toLedger.append(LedgerEntry.Type.CREDIT, amountCents);   // both entries, or neither
            }
        }

        TransferResult result = new TransferResult(amountCents);
        idempotency.put(idempotencyKey, result);
        return result;
    }
}

File: python/lld/money-ledger/digital-wallet/transfer_service.py
class InsufficientFunds(Exception):
    pass


class TransferResult:
    def __init__(self, amount_cents: int):
        self.status = "OK"
        self.amount_cents = amount_cents


class TransferService:
    """Debits one ledger and credits another as a single atomic step. The
    two ledgers' locks are acquired in a FIXED order (by wallet id) so a
    concurrent transfer running the opposite direction can never deadlock
    against this one."""

    def __init__(self, idempotency: "IdempotencyStore"):
        self._idempotency = idempotency

    def transfer(self, from_id, from_ledger, to_id, to_ledger, amount_cents, idempotency_key):
        cached = self._idempotency.get(idempotency_key)
        if cached is not None:
            return cached  # retry — do nothing new

        first, second = (from_ledger, to_ledger) if from_id < to_id else (to_ledger, from_ledger)
        with first.lock, second.lock:
            if from_ledger.get_balance() < amount_cents:
                raise InsufficientFunds()
            from_ledger.append("DEBIT", amount_cents)
            to_ledger.append("CREDIT", amount_cents)  # both entries, or neither

        result = TransferResult(amount_cents)
        self._idempotency.put(idempotency_key, result)
        return result